wagtail/admin/views/pages.py

```diff
@@ -399,11 +399,21 @@ def edit(request, page_id):
 
             # Publish
             if is_publishing:
+                for fn in hooks.get_hooks('before_publish_page'):
+                    result = fn(request, page)
+                    if hasattr(result, 'status_code'):
+                        return result
+
                 revision.publish()
                 # Need to reload the page because the URL may have changed, and we
                 # need the up-to-date URL for the "View Live" button.
                 page = page.specific_class.objects.get(pk=page.pk)
 
+                for fn in hooks.get_hooks('after_publish_page'):
```

Comment:
@allcaps the publish hooks are not tested or documented in this revised PR.

However, maybe we should just get the unpublish hooks merged in and leave the publish hooks for a separate piece of work.

see Matt's comments here
https://github.com/wagtail/wagtail/issues/4590#issuecomment-395806974

If you agree with this, I will get this merged in without the publish hooks.

Reply:
I added:
- Create hooks
- Create hooks tests
- Create hooks docs
- Edit hooks tests
- Edit hooks docs